hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/action/commit/SparkWriteHelper.java

```diff
@@ -58,7 +60,8 @@ public static SparkWriteHelper newInstance() {
       return new Tuple2<>(key, record);
     }).reduceByKey((rec1, rec2) -> {
       @SuppressWarnings("unchecked")
-      T reducedData = (T) rec1.getData().preCombine(rec2.getData());
+      T reducedData = schema != null && !schema.get().isEmpty() ? (T) rec1.getData().preCombine(rec2.getData(), new Schema.Parser().parse(schema.get()))
```

Comment:
Ah another issue here. This can potentially break old payload classes. i.e those classes that don't have precombine(,schema) and be non-backwards compatible.

Reply:
Um.. but we add this in `HoodieRecordPayload`.
```
default T preCombine(T another, Schema schema) throws IOException {
    return preCombine(another);
  }
```
which means the old payload classes will just ignore schema(because they don't need). Only the classes that I just added in this PR will  take advantage of it.
But one problem is that all payload will parse the schema every record whether it needs or not.It will affect performance.

Reply:
> parsing the schema every record will give us performance issues.

Hmm.. Only happened when num of Records with same HoodieKey >2

Reply:
>Only happened when num of Records with same HoodieKey >2

Sorry not following this.